### DISCUSSION TITLE
Expected behavior of copying assets with localized project

### DISCUSSION THREAD

**Discussion by shamoon** [stephanrauh/ngx-extended-pdf-viewer#1978] [Q&A]
Hi there, thank you for this project, I think it's great. I recently went ahead and migrated a project I work on to using this (btw, ng2-pdf-viewer has seen such limited updates recently, you may want to remove it from your docs...) and had a pretty easy time doing that, again, the project works great.

However, when it came time to test this out my builds were taking way longer and ballooned in size from ~100mb to 800mb. After investigating a bit I've discovered that the `assets` dir and all of its contents are being copied dozens of times into the respective localized folders my project builds, in other words I end up with:
- frontend/af-ZA/assets/[all assets]
- frontend/ar-AR/assets/[all assets]
- etc.

I cant say that I have a very deep understanding of the angular build config, so my question is whether this a) is expected, and b) can be avoided? I took a look at https://github.com/stephanrauh/ngx-extended-pdf-viewer/discussions/852 and will implement that but should I still be getting these files repeated under each locale?

Thanks again for any guidance

**Comment by stephanrauh** [comment]
That's weird. The entire showcase - containing both the `assets` and the `bleeding-edge` branch without any optimization - weighs in at "merely" 50 mb. I also have a hard time understanding how Angular can build a recursive folder structure. Maybe you've got a duplicate entry in the `angular.json`?

Create a new project using ng new and ng add ngx-extended-pdf-viewer, as described here: https://pdfviewer.net/extended-pdf-viewer/getting-started. Getting it up and running should be a matter of less than five minutes. After confirming your greenfield project works, compare the `angular.json` of both projects. Usually this helps you to find the issue quickly.

**Reply by shamoon** [reply]
Sorry I forgot to mention I have the `localize` option set. The more I dig around, the more I think this is expected with that. My project has almost 20 languages and the entire locale folder is getting copied into each. It seems like what I really want is each pdf viewer locale in a separate localization folder, but Im not even sure thats achievable and apparently it's discouraged to have a shared assets folder.

I'll keep playing with it, appreciate your thoughts.

**Reply by stephanrauh** [reply]
The dist folder should look like so:
<img width="667" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/4ba87c59-4564-4310-aa64-f12ab6ce5330">
Every language folder contains a single small file.

**Reply by shamoon** [reply]
Yea, with the default angular settings mine looks like this:
<img width="304" alt="Screenshot 2023-11-28 at 11 48 07 PM" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/4887959/55834922-6dfe-4ba6-a372-d56ed7807462"> <img width="321" alt="Screenshot 2023-11-28 at 11 48 37 PM" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/4887959/e6669231-9209-4ef4-bdc8-39686d8db621">

and so on for every language in my app

**Reply by stephanrauh** [reply]
Where do the *.br and *.gz files come from? Is there some post-processing generating these files?

**Reply by shamoon** [reply]
Yea those are artifacts from the build process, here's running just the ng build, again, its the duplication of all the pdf viewer assets inside each of my app's locale folders that makes things add up.

<img width="367" alt="Screenshot 2023-11-29 at 12 27 40 AM" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/4887959/beeb6007-7557-449f-9fba-4c324525b7fe">
<img width="377" alt="Screenshot 2023-11-29 at 12 27 27 AM" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/4887959/e0b1b69d-7d95-4af0-92ea-9e5df214c593">

**Reply by shamoon** [reply]
It's actually the JS assets that are taking up more space, but yea same principle.

**Reply by shamoon** [reply]
I do think this is essentially ‘expected’ it’s just kind of a problem given the size of all the files. But I’m not sure there’s any actual solution, it’s just kinda what angular does, duplicating assets per locale

https://stackoverflow.com/questions/68470821/how-to-create-a-common-assets-folder-for-multiple-locale-in-angular
https://github.com/angular/angular-cli/issues/18414

**Reply by stephanrauh** [reply]
It seems you're using a non-standard configuration. By default, every language folder contains the entire application. When I add ngx-extended-pdf-viewer to the demo you can download here https://angular.io/guide/i18n-example, the `dist` folder looks like so:

<img width="313" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/f3901ee8-6c8b-4aed-b320-b09f12ebd683">

**Reply by shamoon** [reply]
But thats the same thing right? In each of those folders:
dist/browser/en-US/assets/
dist/browser/fr/assets/

you have all of the assets, so the entire /locale/ folder, all of the JS files are duplicated in each location.

Again, Im not suggesting this project is doing anything wrong, and it seems this is just how Angular works but it does make the overall releases very heavy as the assets are duplicated many times.

**Reply by stephanrauh** [reply]
If your top-level folders `af-ZA` and `ar-AR` also contain the main bundle and the `index.html`, you're right, your screenshots show the same behavior as my screenshots. For some reason, I believed you've got only one `index.html` and a top-level `assets` folder that contains many nested `assets` folder.

BTW, I'm not sure I'd call `@angular/localize` "the" standard. It's a fascinating approach that has many advantages, but I've never seen it in the wild yet. That doesn't mean much: I've been merely on two or three projects with i18n. That's not enough to call me an expert. Nonetheless, these projects used ngx-translate (http://www.ngx-translate.com/). That's a less ambitious project. It trades performance for flexibility. In particular, it allows you to change the language at runtime. In contrast, `@angular/localize`requires reloading the application. Maybe that's an option for you.

**Reply by shamoon** [reply]
Yea, they each contain the index.html, main.js, etc., its the whole app duplicated 😑
<img width="248" alt="Screenshot 2023-11-29 at 4 09 35 PM" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/4887959/7ee96022-0c38-4685-b027-7df4291c1caa">

Ha, yea I guess I think of @angular packages as standard, but cool, I havent heard of that other project (it's probably too big of a lift for us at this point). Mostly, I think my project suffers this issue more than most because it has a lot of languages.

Thanks again for your time (and the project!).
